File: tests/csharp/CommentsStrings.cs
// CommentsStrings.cs — verifies that fake declarations inside comments and
// string literals of every flavour are silently ignored.

// class FakeLineCommentClass { void FakeMethod() {} }
// interface IFakeLine { int Compute(); }
// const int FAKE_LINE_CONST = 99;

/* class FakeBlockClass { public void Go() {} } */
/*
 * Multi-line block comment containing fake signatures:
 * public struct FakeStruct { int X; }
 * public enum FakeEnum { A, B }
 * public const double FakeConst = 3.14;
 */

/** <summary>class XmlDocFake { }</summary> */

namespace Demo.CommentsStrings
{
    /// <summary>
    /// This XML doc comment must NOT produce any signatures:
    /// <code>
    /// class DocFakeClass { void DocFakeMethod() {} }
    /// interface IDocFake { string Get(); }
    /// </code>
    /// </summary>
    public class RealClass
    {
        // Ordinary string — fake decls inside must be ignored.
        private const string HelpText =
            "class NotReal { void AlsoNotReal() {} } const int Nope = 0;";

        // Verbatim string — fake decls must also be ignored.
        private static readonly string SqlQuery =
            @"SELECT class, interface FROM schema WHERE enum = 'record'
              AND struct > 0";

        // Triple-quoted (C# 11 raw) string literal with fake decls.
        private static readonly string RawLiteral = """
            class InsideRaw { public void Method() {} }
            interface IInsideRaw { void Foo(); }
            const int RAW_CONST = 42;
            """;

        // Interpolated string — the expressions are real code but the text isn't.
        private string _tag = "real";
        public string Describe(int n) =>
            $"class count: {n} interface count: {n - 1} struct count: 0";

        /// <summary>A real method — it must appear in output.</summary>
        public void RealMethod() { }

        /// <summary>Real constant.</summary>
        public const int RealConst = 7;
    }

    // Line-comment-terminated interface — only the real interface below counts.
    // interface ILineFake { void Phantom(); }

    /// <summary>Real interface.</summary>
    public interface IRealInterface
    {
        /// <summary>
        /// enum FakeInDoc { X }  — must be ignored (inside XML doc)
        /// </summary>
        void RealWork();
        int RealCompute(double value);
    }

    public enum RealEnum
    {
        First,
        Second,
        Third
    }

    public struct RealStruct
    {
        public int Value;
        public RealStruct(int value) { Value = value; }

        /* struct FakeInsideBlockComment { } */
        public override string ToString() => Value.ToString();
    }

    public static class StringHelper
    {
        public static string Repeat(string s, int times) =>
            string.Concat(Enumerable.Repeat(s, times));

        public static bool Contains(string haystack, string needle) =>
            haystack.Contains(needle, StringComparison.OrdinalIgnoreCase);
    }
}

File: tests/go/comments_strings.go
// Package comments_strings verifies that fake declarations inside comments and
// string/raw-string literals are NOT extracted by the signatures CLI.
//
// The following lines inside this block comment must be ignored:
// func fakeInLineComment(x int) int
// type FakeCommentType struct
// const FakeConst = 99
package commentsstrings

// This is a real constant.
const RealConst = 42

// SomeFunc has a comment above it that mentions func fake1(a, b string) bool
// which must not appear in the output.
func SomeFunc(a, b string) bool {
	// Inside the body: func innerFake() -- must be ignored (inside function body)
	_ = "func fakeInDoubleQuoteString(x int) int"
	_ = `func fakeInRawString(y float64) float64
type FakeRawType struct {
	const FakeRawConst = 0
}`
	return false
}

/* Block comment containing:
   func fakeInBlockComment(z int) {}
   type FakeBlock struct { X int }
   const FakeBlockConst = "hello"
*/

// RealType is a real struct declaration.
type RealType struct {
	Name  string
	Value int
}

// FakeMethodComment demonstrates that a method reference inside a comment is ignored:
// func (r RealType) FakeCommentMethod() error
func (r RealType) RealMethod() error {
	return nil
}

// multiLineStr holds a raw string with embedded fake declarations.
var multiLineStr = `
func notAFunc(a int) {
}

type notAType struct {
	X int
}

const notAConst = 7
`

// anotherFn has a double-quoted string with fake Go code.
func anotherFn() string {
	return "type Ghost struct { func fake() }"
}

// RealInterface is a real interface.
type RealInterface interface {
	Do() error
	Undo() error
}
